browser/misc_metrics/vertical_tab_metrics.h

```diff
@@ -0,0 +1,88 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_BROWSER_MISC_METRICS_VERTICAL_TAB_METRICS_H_
+#define BRAVE_BROWSER_MISC_METRICS_VERTICAL_TAB_METRICS_H_
+
+#include <memory>
+
+#include "base/containers/flat_map.h"
+#include "base/memory/raw_ptr.h"
+#include "brave/components/time_period_storage/weekly_storage.h"
+#include "chrome/browser/ui/browser_list_observer.h"
+#include "chrome/browser/ui/tabs/tab_strip_model_observer.h"
+#include "components/prefs/pref_change_registrar.h"
+#include "components/sessions/core/session_id.h"
+
+class Browser;
+class PrefRegistrySimple;
+class PrefService;
+class TabStripModel;
+
+namespace misc_metrics {
+
+extern const char kVerticalOpenTabsHistogramName[];
+extern const char kVerticalGroupTabsHistogramName[];
+extern const char kVerticalPinnedTabsHistogramName[];
+
+enum class TabCountType { kOpen, kGroup, kPinned };
+
+const TabCountType kAllTabCountTypes[] = {
+    TabCountType::kOpen, TabCountType::kGroup, TabCountType::kPinned};
+
+class VerticalTabBrowserMetrics : public TabStripModelObserver {
+ public:
+  explicit VerticalTabBrowserMetrics(PrefService* profile_prefs,
+                                     base::RepeatingClosure change_callback);
+  ~VerticalTabBrowserMetrics() override;
+
+  VerticalTabBrowserMetrics(const VerticalTabBrowserMetrics&) = delete;
+  VerticalTabBrowserMetrics& operator=(const VerticalTabBrowserMetrics&) =
+      delete;
+
+  void OnTabStripModelChanged(
+      TabStripModel* tab_strip_model,
+      const TabStripModelChange& change,
+      const TabStripSelectionChange& selection) override;
+
+  bool IsVerticalTabsEnabled() const;
+  size_t GetTabCount(TabCountType count_type) const;
+
+ private:
+  void UpdateEnabledStatus();
+
+  bool vertical_tabs_enabled_;
+  raw_ptr<PrefService> profile_prefs_;
+  PrefChangeRegistrar pref_change_registrar_;
+  base::flat_map<TabCountType, size_t> counts_;
```

Comment:
nit: FixedFlatMap or a simple std::array could be better.

Reply:
unfortunately `base::FixedFlatMap` is immutable, and `std::array` is less ergonomic since we would have to static_cast the enum class to a size_t upon each access of the array.

Reply:
Good point!